qubes-gui-gntalloc/src/lib.rs

```diff
@@ -0,0 +1,272 @@
+//! Grant-table manipulation code
+
+use std::io;
+use std::os::unix::io::AsRawFd as _;
+use std::rc::{Rc, Weak};
+
+type DomID = u16;
+
+/// A GUI agent instance
+pub struct Agent {
+    alloc: Rc<std::fs::File>,
+    peer: DomID,
+}
+
+/// A window sent to the GUI daemon
+pub struct Buffer {
+    inner: Vec<u64>,
+    alloc: Weak<std::fs::File>,
+    ptr: *mut libc::c_void,
+    offset: u64,
+    dimensions: dimensions::WindowDimensions,
+}
+
+mod dimensions {
+    use std::io;
+    use std::mem::size_of;
+    pub(super) struct WindowDimensions {
+        width: u32,
+        height: u32,
+    }
+
+    impl WindowDimensions {
+        pub fn new(width: u32, height: u32) -> io::Result<Self> {
+            let _: [u8; 0] = [0u8; (size_of::<u32>() > size_of::<usize>()) as usize
+                + ((u32::MAX / 4) / qubes_gui::MAX_WINDOW_WIDTH <= qubes_gui::MAX_WINDOW_HEIGHT)
+                    as usize];
+            if width > qubes_gui::MAX_WINDOW_WIDTH || height > qubes_gui::MAX_WINDOW_HEIGHT {
+                return Err(io::Error::new(
+                    io::ErrorKind::InvalidInput,
+                    format!(
+                        "Window dimensions {}x{} too large (limit is {}x{})",
+                        width,
+                        height,
+                        qubes_gui::MAX_WINDOW_WIDTH,
+                        qubes_gui::MAX_WINDOW_HEIGHT
+                    ),
+                ));
+            }
+            if width == 0 || height == 0 {
+                return Err(io::Error::new(
+                    io::ErrorKind::InvalidInput,
+                    format!(
+                        "Window dimensions {}x{} not valid: neither width nor height may be zero",
+                        width, height,
+                    ),
+                ));
+            }
+            Ok(Self { width, height })
+        }
+
+        pub fn buffer_size(&self) -> usize {
+            assert!(self.width <= qubes_gui::MAX_WINDOW_WIDTH, "checked earlier");
```

Comment:
Type safety means we can easily prove to ourselves that this is an invariant the type will satisfy. As such I don't think it's necessary to check this in release mode:

```suggestion
            debug_assert!(self.width <= qubes_gui::MAX_WINDOW_WIDTH, "checked earlier");
```